src/VisualStudio/Core/Impl/CodeModel/ProjectCodeModelFactory.cs

```diff
@@ -32,15 +38,91 @@ internal sealed class ProjectCodeModelFactory : IProjectCodeModelFactory, IDispo
+
+            // Queue up notifications we hear about docs changing.  that way we don't have to fire events multiple times
+            // for the same documents.  Once enough time has passed, take the documents that were changed and run
+            // through them, firing their latest events.
+            _documentsToFireEventsFor = new AsyncBatchingWorkQueue<DocumentId>(
+                TimeSpan.FromMilliseconds(visualStudioWorkspace.Options.GetOption(InternalSolutionCrawlerOptions.AllFilesWorkerBackOffTimeSpanInMS)),
+                ProcessNextDocumentBatchAsync,
+                // We only care about unique doc-ids, so pass in this comparer to collapse streams of changes for a
+                // single document down to one notification.
+                EqualityComparer<DocumentId>.Default,
+                _listener,
+                _cancellationTokenSource.Token);
+
+            _visualStudioWorkspace.WorkspaceChanged += OnWorkspaceChanged;
+        }
+
+        private System.Threading.Tasks.Task ProcessNextDocumentBatchAsync(
+            ImmutableArray<DocumentId> documentIds, CancellationToken cancellationToken)
+        {
+            foreach (var documentId in documentIds)
+            {
+                // Now, enqueue foreground work to actually process these documents in a serialized and incremental
+                // fashion.  FireEventsForDocument will actually limit how much time it spends firing events so that it
+                // doesn't saturate  the UI thread.
+                _notificationService.RegisterNotification(
+                    () => FireEventsForDocument(documentId),
+                    _listener.BeginAsyncOperation("CodeModelEvent"),
+                    cancellationToken);
+            }
+
+            return System.Threading.Tasks.Task.CompletedTask;
+
+            bool FireEventsForDocument(DocumentId documentId)
+            {
+                // If we've been asked to shutdown, don't bother reporting any more events.
+                if (_cancellationTokenSource.IsCancellationRequested)
+                    return false;
+
+                var projectCodeModel = this.TryGetProjectCodeModel(documentId.ProjectId);
+                if (projectCodeModel == null)
+                    return false;
+
+                var filename = _visualStudioWorkspace.GetFilePath(documentId);
+                if (filename == null)
+                    return false;
+
+                if (!projectCodeModel.TryGetCachedFileCodeModel(filename, out var fileCodeModelHandle))
+                    return false;
+
+                var codeModel = fileCodeModelHandle.Object;
+                return codeModel.FireEvents();
+            }
+        }
+
+        private void OnWorkspaceChanged(object sender, WorkspaceChangeEventArgs e)
+        {
+            if (e.Kind == WorkspaceChangeKind.DocumentChanged || e.Kind == WorkspaceChangeKind.DocumentRemoved)
```

Comment:
@jasonmalinowski Is it possible that some workspace operations aggregate a change in document with other changes to the project, for example, and send WorkspaceChangeKind.ProjectChanged and no DocumentChanged?

Reply:
oh gosh. i hope not.  it's unclear to me how features would be able to manage things effectively in that case :-/

Reply:
Absolutely; if a bunch of documents are all changing and we raise just a single event you get a single ProjectChanged.

Reply:
(and if two documents change in different projects simultaneously...you might have a single SolutionChanged)

Reply:
oh god.  so how does solution crawler work?  @mavasani How do you know what sort of indiviudal analysis operations to run whne you hear about bulk changes from the Workspace?

Reply:
WorkCoordinator operates on ProjectChanges for ProjectChanged events: http://sourceroslyn.io/#Microsoft.CodeAnalysis.Features/SolutionCrawler/WorkCoordinator.cs,540

Reply:
Thanks.  have switched over.